### PR TITLE
Add a NewOptimizer function according to string flags

### LINKED ISSUE #1703: add a NewOptimizer function according to string flags
Now the optimizer flags are parsed in master into a string pair, and then pass to PS in Go. Please refer to this [PR](https://github.com/sql-machine-learning/elasticdl/pull/1695).

We should add a `NewOptimizer` function to accept the string pair, and returns a Optimizer instance in [optimizer.go](https://github.com/sql-machine-learning/elasticdl/blob/develop/elasticdl/pkg/ps/optimizer.go)

### PR DESCRIPTION
**Pull request by mhaoli**
Fixes #1703 . Changes include:
* `parseOptArgs` parses optimizer arguments according to optimizer type.
* `NewOptimizer` creates optimizer according to optimizer type and arguments parsed by `parseOptArgs`.

### CODE REVIEW

File: elasticdl/pkg/ps/optimizer.go [hunk 1]
```diff
@@ -63,6 +67,11 @@ type AdamOptimizer struct {
 	maxSquare *Parameter
 }
 
+// GetLR returns learning rate Adam
```
Comment (skydoorkai) [inline elasticdl/pkg/ps/optimizer.go]: // GetLR returns learning rate (delete Adam)

File: elasticdl/pkg/ps/optimizer.go [hunk 2]
```diff
@@ -130,11 +139,106 @@ func (opt *AdamOptimizer) InitNonEmbeddingParam(name string, dim []int64) {
...
+	}
+
+	// check argument names
+	for _, argName := range optArgumentsMap[optType] {
+		if _, ok := argsMap[argName]; !ok {
+			return nil, fmt.Errorf("Args passed to ps should contain %s", argName)
+		}
+	}
+	if len(argsMap) != len(optArgumentsMap[optType]) {
+		return nil, fmt.Errorf("Args passed to ps contain redundant items: %v", argsMap)
```
Comment (skydoorkai) [inline elasticdl/pkg/ps/optimizer.go:179]: The other case is that args passed to ps do not have all the required items. May change to:
"Args do not match %v optimizer", optType

Comment (mhaoli) [reply elasticdl/pkg/ps/optimizer.go:179]: Above code checked that args passed to ps have all the required items. Please refer to 

https://github.com/sql-machine-learning/elasticdl/pull/1717/files/eb25d2762e8889c80d9afebbd1dec9f5b6338cd5#diff-245e29d9f4a07ba101d299a9f3e5356bR172-R177

Comment (skydoorkai) [reply elasticdl/pkg/ps/optimizer.go:179]: I see.

File: elasticdl/pkg/ps/optimizer.go (hunk 1 above)
Comment (mhaoli) [reply elasticdl/pkg/ps/optimizer.go]: Done

**Comment by QiJune** [comment]
This PR has some conflicts to be resolved.

**Comment by mhaoli** [comment]
> This PR has some conflicts to be resolved.

Done.

File: elasticdl/pkg/ps/server.go [hunk 3]
```diff
@@ -23,11 +23,17 @@ type Server struct {
...
+func NewServer(ID int, optType string, optArgs string) *Server {
+	var ps Server
+	ps.Param = NewParameter()
+	var err error
+	ps.Opt, err = NewOptimizer(optType, optArgs)
+	if err != nil {
+		log.Fatalf("failed to create PS server: %v", err)
+	}
+	ps.ID = ID
+	ps.lock = sync.Mutex{}
```
Comment (QiJune) [inline elasticdl/pkg/ps/server.go]: We could remove this line. The zero value for a Mutex is an unlocked mutex. Please refer to https://stackoverflow.com/questions/45744165/do-mutexes-need-initialization-in-go

**Review by QiJune — APPROVED** [review]

File: elasticdl/pkg/ps/server.go (hunk 3 above)
Comment (mhaoli) [reply elasticdl/pkg/ps/server.go]: Done

**Pull request merged**